lib/snitch-core/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+    pub const DEFAULT: Self = Self {
+        storage: heapless::Deque::new(),
+        earliest_ena: 0,
+        insert_state: InsertState::Collecting,
+        our_task_id: 0,
+    };
+
+    /// Sets up the `Store` with data from the environment. Must be called once
+    /// before other functions.
+    ///
+    /// This stores the bits of the current task's `TaskId` (`tid`) and the
+    /// current system `timestamp`, to be used in generating loss records.
+    pub fn initialize(&mut self, tid: u16, timestamp: u64) {
+        self.our_task_id = tid;
+        // If the queue has never been touched, insert our "arbitrary data loss"
+        // record into the stream to consume ENA 0.
+        if self.earliest_ena == 0 && self.storage.is_empty() {
+            self.insert_impl(
+                self.our_task_id,
+                timestamp,
+                &[0xA1, 0x64, 0x6C, 0x6F, 0x73, 0x74, 0xF6],
+            );
+        }
+    }
+
+    /// Internal check for whether `initialize` has been called.
+    fn initialized(&self) -> bool {
+        self.earliest_ena != 0 || !self.storage.is_empty()
+    }
+
+    /// Inserts a record, or records it as lost.
+    ///
+    /// This attempts to record a record from `sender` at `timestamp` containing
+    /// `data`. If the record can't be stored, this records it as lost. No
+    /// indication is given to the caller, since the caller can't really do
+    /// anything anyway.
+    ///
+    /// If we're already in a `Losing` situation, but data has been flushed such
+    /// that we might yet be able to make forward progress, this will attempt to
+    /// recover before inserting this record.
+    pub fn insert(&mut self, sender: u16, timestamp: u64, data: &[u8]) {
+        debug_assert!(self.initialized());
+        self.insert_impl(sender, timestamp, data)
+    }
+
+    /// Iterates over the entire current contents of the store.
+    ///
+    /// This will produce `Record`s in the order they were received, with
+    /// ascending ENA values.
+    pub fn iter_contents(&self) -> impl Iterator<Item = Record<'_>> + '_ {
+        let mut slices = self.storage.as_slices();
+        let mut next_ena = self.earliest_ena;
+        core::iter::from_fn(move || {
+            let len = u16::from_le_bytes(take_array(&mut slices)?) as usize;
+            let tid = u16::from_le_bytes(take_array(&mut slices)?);
+            let timestamp = u64::from_le_bytes(take_array(&mut slices)?);
+            let slices = take_slice(&mut slices, len)?;
+            let ena = next_ena;
+            next_ena += 1;
+            Some(Record {
+                ena,
+                tid,
+                timestamp,
+                slices,
+            })
+        })
+    }
+
+    /// Reads records starting from `first_ena`, inclusive.
+    pub fn read_from(
+        &mut self,
+        first_ena: u64,
+    ) -> impl Iterator<Item = Record<'_>> + '_ {
+        // Attempt to recover and insert a loss record if necessary.
```

Comment:
Why does this function also attempt to recover, versus `iter_contents` which is immutable?

Reply:
I actually switched the behavior of `iter_contents` after writing tests against it, it turns out to be rather annoying to have it not attempt recovery.